### PR TITLE
CASEFLOW-126 #done Copy AOD state to CAVC Remand Appeal

### PR DESCRIPTION
**Pull request by yoomlam**
Resolves [CASEFLOW-126](https://vajira.max.gov/browse/CASEFLOW-126)

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] Verify the CAVC Remand is AOD if the original appeal was AOD
How we determine AOD:
- [ ] `aod_based_on_age` is true on the appeal (set by a job, should be set automatically upon creation)
- [ ] The claimant's dob is over 75 years ago (should have the same claimant, automatically checked in the code)
- [ ] The claimant has a granted age base aod motion (AOD currently only applies at the combined claimant and appeal level; create a new aod motion for this appeal if one existed for the OG)

### Testing Plan
Refer to Rails commands provided in #15469.

```ruby
# Create an active "source appeal" with AOD due to age
# Also try this with another appeal that is not AOD (leave out `:advanced_on_docket_due_to_age`).
appeal = FactoryBot.create(:appeal, :active, :advanced_on_docket_due_to_age)
appeal.aod_based_on_age
=> true
appeal.aod?
=> true

decision_issues = FactoryBot.create_list(
        :decision_issue,
        3,
        :rating,
        decision_review: appeal,
        disposition: "denied",
        description: "Decision issue description",
        decision_text: "decision issue"
      )
params = {
        created_by: User.last,
        updated_by: User.last,
        appeal: appeal,
        cavc_docket_number: "123-1234567",
        represented_by_attorney: true,
        cavc_judge_full_name: Constants::CAVC_JUDGE_FULL_NAMES.first,
        cavc_decision_type: Constants::CAVC_DECISION_TYPES.keys.first,
        remand_subtype: Constants::CAVC_REMAND_SUBTYPES.keys.first,
        decision_date: 5.days.ago.to_date,
        judgement_date: 4.days.ago.to_date,
        mandate_date: 3.days.ago.to_date,
        decision_issue_ids: decision_issues.map(&:id),
        instructions: "Instructions!"
      }
cavc_remand = CavcRemand.create!(params)
cavc_appeal = Appeal.court_remand.where(stream_docket_number: cavc_remand.appeal.docket_number).order(:id).last
cavc_appeal.aod_based_on_age
=> true
cavc_appeal.aod?
=> true
```

```ruby
# Create an active "source appeal" with AOD due to AOD motion
appeal = FactoryBot.create(:appeal, :active, :advanced_on_docket_due_to_motion)
appeal.aod?
=> true

decision_issues = ... # same as above
params  = ... # same as above
cavc_remand = CavcRemand.create!(params)
cavc_appeal = Appeal.court_remand.where(stream_docket_number: cavc_remand.appeal.docket_number).order(:id).last
cavc_appeal.aod?
=> true
cavc_appeal.aod_based_on_age
=> false

# Verify that there are new AdvanceOnDocketMotion -- see code in RSpec tests
```

### CODE REVIEW

File: spec/factories/appeal.rb [hunk 1]
```diff
@@ -153,8 +153,13 @@
     end
 
     trait :advanced_on_docket_due_to_age do
-      after(:create) do |appeal, _evaluator|
-        appeal.claimants = [create(:claimant, :advanced_on_docket_due_to_age, decision_review: appeal)]
+      # set claimant.decision_review to nil so that it isn't created by the Claimant factorybot
+      claimants { [create(:claimant, :advanced_on_docket_due_to_age, decision_review: nil)] }
+    end
```
Comment (yoomlam) [inline spec/factories/appeal.rb:158]: `claimant.decision_review` will point to the appeal being created by this factory.

File: spec/factories/appeal.rb [hunk 2]
```diff
@@ -153,8 +153,13 @@
...
+
+    trait :active do
+      before(:create) do |appeal, _evaluator|
+        RootTask.find_or_create_by!(appeal: appeal, assigned_to: Bva.singleton)
```
Comment (yoomlam) [inline spec/factories/appeal.rb:162]: Based on the implementation of `Appeal.active?`. `set_appeal_age_aod_job_spec.rb` differentiated between active and inactive appeals.

File: spec/models/cavc_remand_spec.rb [hunk 3]
```diff
@@ -52,10 +52,50 @@
       params.each_key { |key| expect(subject.send(key)).to eq params[key] }
     end
 
+    def last_cavc_appeal(cavc_remand)
+      # To-do: remove assumption that the last appeal with the same document number is the created appeal
+      # We should be able to get the cavc_appeal from cavc_remand in case there are multiple found
+      Appeal.court_remand.where(stream_docket_number: cavc_remand.appeal.docket_number).order(:id).last
+    end
```
Comment (yoomlam) [inline spec/models/cavc_remand_spec.rb:59]: Hunter will write up a ticket for this `To-do` -- [Slack](https://dsva.slack.com/archives/CJL810329/p1608564559171000?thread_ts=1608564273.168800&cid=CJL810329)

File: spec/models/cavc_remand_spec.rb [hunk 4]
```diff
@@ -52,10 +52,50 @@
...
+
     it "creates the new court_remand cavc stream" do
       expect(Appeal.court_remand.where(stream_docket_number: appeal.docket_number).count).to eq(0)
+      expect(appeal.aod_based_on_age).not_to be true
       expect { subject }.not_to raise_error
       expect(Appeal.court_remand.where(stream_docket_number: appeal.docket_number).count).to eq(1)
+      expect(last_cavc_appeal(subject).aod_based_on_age).not_to be true
```
Comment (yoomlam) [inline spec/models/cavc_remand_spec.rb:66]: Verify that `aod_based_on_age` isn't true when source appeal is not true.

**Review by yoomlam — COMMENTED** [review]
Commentary

File: app/models/advance_on_docket_motion.rb [hunk 5]
```diff
@@ -59,6 +60,18 @@ def create_or_update_by_appeal(appeal, attrs)
       end
     end
 
+    def copy_to_appeal(src_appeal, dst_appeal)
+      person = src_appeal.claimant.person
+      fail "Claimants on appeals are expected to be the same" unless person == dst_appeal.claimant.person
+
+      where(person_id: person, appeal: src_appeal).map do |aod_motion|
```
Comment (hschallhorn) [inline app/models/advance_on_docket_motion.rb:67]: do we want to add `.granted` here? or do we want to copy over denied motions as well?

File: spec/models/cavc_remand_spec.rb [hunk 6]
```diff
@@ -27,7 +27,7 @@
       )
     end
     let(:decision_issue_ids) { decision_issues.map(&:id) }
-    let(:instructions) { "Intructions!" }
+    let(:instructions) { "Instructions!" }
```
Comment (hschallhorn) [inline spec/models/cavc_remand_spec.rb:30]: Don't even need git blame to know who this was

**Review by hschallhorn — COMMENTED** [review]
One tiny question, running through testing now!

**Review by hschallhorn — COMMENTED** [review]
- [ ] AOD due to claimant age
```ruby
appeal = BvaDispatchTask.completed.first.appeal
appeal.aod?
=> true
# oh, hm, guess I don't need to do anything! Let's find out why right quick
appeal.aod_based_on_age
=> true
appeal.claimant.person.date_of_birth
=> Thu, 14 Dec 1944
# 76 years old, cool!
appeal.claimant.advanced_on_docket_motion_granted?(appeal)
=> false
# Alright, let's intake a cavc remand and ensure the new appeal is aod
```
OG appeal
Screen Shot 2020-12-22 at 1 27 27 PM
CAVC appeal
Screen Shot 2020-12-22 at 1 28 00 PM

- [ ] AOD based on granted motion
```ruby
appeal = BvaDispatchTask.completed.last.appeal
appeal.aod?
=> false 
# perrrrrrfect, let's grant them a motion
AdvanceOnDocketMotion.create_or_update_by_appeal(
  appeal,
  reason: AdvanceOnDocketMotion.reasons["other"],
  granted: true,
  user_id: User.first.id
)
appeal.aod? 
=> true
# Alright, let's intake a cavc remand and ensure the new appeal is aod
```
OG appeal
Screen Shot 2020-12-22 at 1 41 30 PM
CAVC appeal
Screen Shot 2020-12-22 at 1 41 46 PM
🤔  Seems this may not be working for aod motions, digging

File: app/models/advance_on_docket_motion.rb (hunk 5 above)
Comment (yoomlam) [reply app/models/advance_on_docket_motion.rb:67]: Hmmm. Not sure. We can always look at the source appeal for denied motions, but it _might_ be convenient (for the Reporting Team) to know denied motions.  Hmmm. ... but that may suggest that the motion was denied on this CAVC Remand appeal, so let's not copy over denied motions.

**Review by hschallhorn — APPROVED** [review]
Lovely!

File: app/models/advance_on_docket_motion.rb (hunk 5 above)
Comment (yoomlam) [reply app/models/advance_on_docket_motion.rb:67]: Renaming `copy_to_appeal` to `copy_granted_motions_to_appeal`

**Pull request merged**